arangod/Replication2/ReplicatedState/ReplicatedState.tpp

```diff
@@ -30,33 +31,692 @@
+  // TODO Pass the stream during construction already, and delete the
+  //      "setStream" method; after that, the leader state implementation can
+  //      also really rely on the stream being there.
+  leaderState->setStream(stream);
+  auto& manager =
+      guard->_currentManager
+          .template emplace<std::shared_ptr<LeaderStateManager<S>>>(
+              std::make_shared<LeaderStateManager<S>>(
+                  _loggerContext.template with<logContextKeyStateRole>(
+                      static_strings::StringLeader),
+                  _metrics, std::move(leaderState), std::move(stream)));
+
+  // temporary hack: post on the scheduler to avoid deadlocks with the log
+  auto& scheduler = *SchedulerFeature::SCHEDULER;
+  scheduler.queue(RequestLane::CLUSTER_INTERNAL,
+                  [weak = manager->weak_from_this()]() mutable {
+                    if (auto manager = weak.lock(); manager != nullptr) {
+                      manager->recoverEntries();
+                    }
+                  });
+}
+
+template<typename S>
+void ReplicatedStateManager<S>::becomeFollower(
+    std::unique_ptr<replicated_log::IReplicatedLogFollowerMethods> methods) {
+  auto guard = _guarded.getLockedGuard();
+  ADB_PROD_ASSERT(
+      std::holds_alternative<std::shared_ptr<UnconfiguredStateManager<S>>>(
+          guard->_currentManager));
+  auto&& [core, oldMethods] =
+      std::move(*std::get<std::shared_ptr<UnconfiguredStateManager<S>>>(
+                    guard->_currentManager))
+          .resign();
+  ADB_PROD_ASSERT(oldMethods == nullptr);
+
+  auto followerState = _factory->constructFollower(std::move(core));
+  auto stream = std::make_shared<StreamProxy<EntryType, Deserializer>>(
+      std::move(methods));
+  followerState->setStream(stream);
+  auto stateManager = std::make_shared<FollowerStateManager<S>>(
+      _loggerContext.template with<logContextKeyStateRole>(
+          static_strings::StringFollower),
+      _metrics, followerState, std::move(stream));
+  followerState->setStateManager(stateManager);
+  guard->_currentManager
+      .template emplace<std::shared_ptr<FollowerStateManager<S>>>(
+          std::move(stateManager));
+}
+
+template<typename S>
+void ReplicatedStateManager<S>::dropEntries() {
+  ADB_PROD_ASSERT(false);
+}
+
+template<typename S>
+auto ReplicatedStateManager<S>::getStatus() const
+    -> std::optional<StateStatus> {
+  auto guard = _guarded.getLockedGuard();
+  auto status = std::visit(
+      overload{[](auto const& manager) { return manager->getStatus(); }},
+      guard->_currentManager);
+
+  return status;
+}
+
+template<typename S>
+auto ReplicatedStateManager<S>::getFollower() const
+    -> std::shared_ptr<IReplicatedFollowerStateBase> {
+  auto guard = _guarded.getLockedGuard();
+  return std::visit(
+      overload{
+          [](std::shared_ptr<FollowerStateManager<S>> const& manager) {
+            return std::dynamic_pointer_cast<IReplicatedFollowerStateBase>(
```

Comment:
dynamic cast is very expensive. Do you really need it in this file?
Maybe you can use basics::downCast? It also works with shared_ptr

Reply:
`getFollower()` is currently not called at all in any production code, only in tests, and `getLeader()` is called only once in `ensureCollection`. I currently see no need to change the performance, and rather have the tiny safety guarantees a dynamic cast provides.
Also, this code path is subject to change, which is another reason why I currently don't want to bother with micro optimizations here.

Reply:
Why don't you want use downCast if it's possible?

Reply:
The main reason is more clear code.
Now from code you cannot understand needs or not dynamic_cast

Reply:
> Why don't you want use downCast if it's possible?

It also makes dynamic_cast check in maintainer mode
(if it's test only code, it's mean always)

But it also make static_cast which works only for base/derived classes

Reply:
> The main reason is more clear code.
> Now from code you cannot understand needs or not dynamic_cast

I can get behind that.